### ISSUE TITLE
Need a better way to prevent property renaming.

### ISSUE THREAD

**Issue by si-robertson** [google/closure-compiler#953]
Hi guys,

This is an issue that has been bugging me for a while now, and it has raised its head again while I have been working on a new 3D framework thing.

AFAIK, preventing property names from being collapsed when compiling code (advanced optimizations) currently requires those properties to be defined ["like"] ["this"] which is, for lack of a better word, horrid. It goes against the grain when someone is trying to write clean source code.

A specially flagged externs file may be the solution to this but the compiler would have to treat it slightly differently from a regular externs file to avoid all of the warnings related to redefinitions etc.

Here's what I'm thinking. You have a codebase of some kind and a standard externs file for that codebase which reflects the public API. Standard stuff. When the codebase is compiled the externs file is also passed to the compiler via a special flag (not the --externs flag) and that will allow the compiler to determine which property names (and class names) etc can be collapsed.

In a nutshell, we say to the compiler "hey, here's an externs file containing the public API, do not rename any of these things but feel free to rename everything else".

Is that a realistic solution?

**Comment by ChadKillingsworth** [comment]
There is a lot of issues with this. Take a look at #930 and the linked discussions and design docs.

In general, `@export` is what you want to use - we just need to work on it some more.

**Comment by si-robertson** [comment]
`--exports_mode NO_RENAME` looks very good to me.

Is that still in the proposal stage at the moment?

**Comment by ChadKillingsworth** [comment]
Yes, but it's high on my list to work on. I'll be following up on the issue at our biweekly meeting tomorrow.

**Comment by si-robertson** [comment]
Marvelous, thanks Chad.
